### PR TITLE
Flag parameters as sensitive if they could contain the database password

### PR DESCRIPTION
**Pull request by derrabus**
|      Q       |   A
|------------- | -----------
| Type         | feature
| Fixed issues | N/A

#### Summary

PHP 8.2 introduced a `SensitiveParameter` attribute that allows us to flag parameters that should be redacted in stack traces: https://wiki.php.net/rfc/redact_parameters_in_back_traces

This PR uses this attribute on all parameters (that I could find) which could possibly contain the database password (or an SSL key), which also includes database URLs.

### CODE REVIEW

File: src/Cache/QueryCacheProfile.php [hunk 1]
```diff
@@ -117,9 +117,7 @@ public function getCacheKey()
      */
     public function generateCacheKeys($sql, $params, $types, array $connectionParams = [])
     {
-        if (isset($connectionParams['password'])) {
-            unset($connectionParams['password']);
-        }
+        unset($connectionParams['password'], $connectionParams['url']);
```
Comment (greg0ire) [inline src/Cache/QueryCacheProfile.php]: Should this be in a separate commit? I don't see the link with sensitive parameters. Also, does this have security implications?

Comment (derrabus) [reply src/Cache/QueryCacheProfile.php]: At this point, the URL should be resolved to individual parameters. If it's still in there, it's 100% redundant. But you are right, I don't need this particular change. I'll move it into a separate PR.

Comment (greg0ire) [reply src/Cache/QueryCacheProfile.php]: As you wish, but a separate commit is good enough.

Comment (derrabus) [reply src/Cache/QueryCacheProfile.php]: I've reverted the changes to this file.

File: src/Connection.php [hunk 2]
```diff
@@ -1099,7 +1101,7 @@ public function executeCacheQuery($sql, $params, $types, QueryCacheProfile $qcp)
         }
 
         $connectionParams = $this->params;
-        unset($connectionParams['platform']);
+        unset($connectionParams['platform'], $connectionParams['password'], $connectionParams['url']);
```
Comment (greg0ire) [inline src/Connection.php:1104]: This looks a lot like the piece of code I previously commented.

File: src/Driver.php [hunk 3]
```diff
@@ -7,23 +7,30 @@
 use Doctrine\DBAL\Driver\Exception;
 use Doctrine\DBAL\Platforms\AbstractPlatform;
 use Doctrine\DBAL\Schema\AbstractSchemaManager;
+use SensitiveParameter;
 
 /**
  * Driver interface.
  * Interface that all DBAL drivers must implement.
+ *
+ * @psalm-import-type Params from DriverManager
```
Comment (greg0ire) [inline src/Driver.php:16]: Should static analysis stuff be in a separate commit?

File: src/Driver/Mysqli/Driver.php [hunk 4]
```diff
@@ -32,27 +34,9 @@ public function connect(array $params)
...
-                $flags = $driverOptions[Connection::OPTION_FLAGS];
-                unset($driverOptions[Connection::OPTION_FLAGS]);
-            }
-
-            $preInitializers = $this->withOptions($preInitializers, $driverOptions);
-        }
-
-        $preInitializers  = $this->withSecure($preInitializers, $params);
-        $postInitializers = $this->withCharset($postInitializers, $params);
-
```
Comment (greg0ire) [inline src/Driver/Mysqli/Driver.php:52]: What's the link with this refactoring and marking parameters as sensitive?

File: src/Connection.php (hunk 2 above)
Comment (derrabus) [reply src/Connection.php:1104]: The `$connectionParams` array is passed to `generateCacheKeys()` in the next line, although we don't need those pieces of information to calculate a good cache key. We have to options here:

* Strip sensitive keys from the `$connectionParams` array or
* flag the second parameter of `generateCacheKeys()` as sensitive as well.

I went for the first option.

File: src/Driver.php (hunk 3 above)
Comment (derrabus) [reply src/Driver.php:16]: Do you want me to split this off? I was merely fixing some inconsistencies on the go.

File: src/Driver/Mysqli/Driver.php (hunk 4 above)
Comment (derrabus) [reply src/Driver/Mysqli/Driver.php:52]: I would need to flag the initializers arrays as sensitive because one of them wraps an SSL key. I wanted to avoid that.

File: src/Driver.php (hunk 3 above)
Comment (greg0ire) [reply src/Driver.php:16]: Just did it for you :slightly_smiling_face:

**Review by greg0ire — APPROVED** [review]
